server/src/main/java/org/opensearch/node/remotestore/RemoteStorePinnedTimestampService.java

```diff
@@ -0,0 +1,272 @@
+    }
+
+    private void validateRemoteStoreConfiguration() {
+        final String remoteStoreRepo = settings.get(
+            Node.NODE_ATTRIBUTES.getKey() + RemoteStoreNodeAttribute.REMOTE_STORE_SEGMENT_REPOSITORY_NAME_ATTRIBUTE_KEY
+        );
+        assert remoteStoreRepo != null : "Remote Cluster State repository is not configured";
+        final Repository repository = repositoriesService.get().repository(remoteStoreRepo);
+        assert repository instanceof BlobStoreRepository : "Repository should be instance of BlobStoreRepository";
+        blobStoreRepository = (BlobStoreRepository) repository;
+    }
+
+    private void initializeComponents() {
+        String clusterName = ClusterName.CLUSTER_NAME_SETTING.get(settings).value();
+        blobStoreTransferService = new BlobStoreTransferService(blobStoreRepository.blobStore(), this.threadPool);
+        pinnedTimestampsBlobStore = new RemoteStorePinnedTimestampsBlobStore(
+            blobStoreTransferService,
+            blobStoreRepository,
+            clusterName,
+            this.threadPool,
+            ThreadPool.Names.REMOTE_STATE_READ
+        );
+    }
+
+    private void startAsyncUpdateTask() {
+        asyncUpdatePinnedTimestampTask = new AsyncUpdatePinnedTimestampTask(logger, threadPool, pinnedTimestampsSchedulerInterval, true);
+    }
+
+    /**
+     * Pins a timestamp in the remote store.
+     *
+     * @param timestamp The timestamp to be pinned
+     * @param pinningEntity The entity responsible for pinning the timestamp
+     * @param listener A listener to be notified when the pinning operation completes
+     * @throws IllegalArgumentException If the timestamp is less than the current time minus one second
+     */
+    public void pinTimestamp(long timestamp, String pinningEntity, ActionListener<Void> listener) {
+        // If a caller uses current system time to pin the timestamp, following check will almost always fail.
+        // So, we allow pinning timestamp in the past upto some buffer
+        if (timestamp < TimeUnit.NANOSECONDS.toMillis(System.nanoTime()) - TIMESTAMP_PINNING_PAST_BUFFER_IN_MILLIS) {
+            throw new IllegalArgumentException("Timestamp to be pinned is less than current timestamp");
+        }
+        updatePinning(pinnedTimestamps -> pinnedTimestamps.pin(timestamp, pinningEntity), listener);
+    }
+
+    /**
+     * Unpins a timestamp from the remote store.
+     *
+     * @param timestamp The timestamp to be unpinned
+     * @param pinningEntity The entity responsible for unpinning the timestamp
+     * @param listener A listener to be notified when the unpinning operation completes
+     */
+    public void unpinTimestamp(long timestamp, String pinningEntity, ActionListener<Void> listener) {
+        updatePinning(pinnedTimestamps -> pinnedTimestamps.unpin(timestamp, pinningEntity), listener);
+    }
+
+    private void updatePinning(Consumer<PinnedTimestamps> updateConsumer, ActionListener<Void> listener) {
```

Comment:
is concurrent executing of this method lead to remote pinned timestamp store to get corrupted?

Reply:
Good point! It can, let me change to make it thread-safe.